dipy/io/stateful_tractogram.py

```diff
@@ -0,0 +1,798 @@
+        elif self._space == Space.RASMM:
+            self._rasmm_to_vox()
+
+    def to_voxmm(self):
+        """ Safe function to transform streamlines and update state """
+        if self._space == Space.VOX:
+            self._vox_to_voxmm()
+        elif self._space == Space.RASMM:
+            self._rasmm_to_voxmm()
+
+    def to_rasmm(self):
+        """ Safe function to transform streamlines and update state """
+        if self._space == Space.VOX:
+            self._vox_to_rasmm()
+        elif self._space == Space.VOXMM:
+            self._voxmm_to_rasmm()
+
+    def to_center(self):
+        """ Safe function to shift streamlines so the center of voxel is
+        the origin """
+        if self._shifted_origin:
+            self._shift_voxel_origin()
+
+    def to_corner(self):
+        """ Safe function to shift streamlines so the corner of voxel is
+        the origin """
+        if not self._shifted_origin:
+            self._shift_voxel_origin()
+
+    def compute_bounding_box(self):
+        """ Compute the bounding box of the streamlines in their current state
+
+        Returns
+        -------
+        output : ndarray
+            8 corners of the XYZ aligned box, all zeros if no streamlines
+        """
+        if self._tractogram.streamlines.data.size > 0:
+            bbox_min = np.min(self._tractogram.streamlines.data, axis=0)
+            bbox_max = np.max(self._tractogram.streamlines.data, axis=0)
+
+            return np.asarray(list(product(*zip(bbox_min, bbox_max))))
+
+        return np.zeros((8, 3))
+
+    def is_bbox_in_vox_valid(self):
+        """ Verify that the bounding box is valid in voxel space
+        Will transform the streamlines for OBB, slow for big tractogram
+
+        Returns
+        -------
+        output : bool
+            Are the streamlines within the volume of the associated reference
+        """
+        old_space = deepcopy(self.get_current_space())
+        old_shift = deepcopy(self.get_current_shift())
+
+        # Do to rotation, equivalent of a OBB must be done
+        self.to_vox()
+        self.to_corner()
+        bbox_corners = deepcopy(self.compute_bounding_box())
+
+        is_valid = True
+        if np.any(bbox_corners < 0):
+            logging.error('Voxel space values lower than 0.0')
+            logging.debug(bbox_corners)
+            is_valid = False
+
+        if np.any(bbox_corners[:, 0] > self._dimensions[0]) or \
+                np.any(bbox_corners[:, 1] > self._dimensions[1]) or \
+                np.any(bbox_corners[:, 2] > self._dimensions[2]):
+            logging.error('Voxel space values higher than dimensions')
+            logging.debug(bbox_corners)
+            is_valid = False
+
+        if old_space == Space.RASMM:
+            self.to_rasmm()
+        elif old_space == Space.VOXMM:
+            self.to_voxmm()
+
+        if not old_shift:
+            self.to_center()
+
+        return is_valid
+
+    def remove_invalid_streamlines(self):
```

Comment:
@frheault what is the exact purpose of this method?

Reply:
Sometime directly after generation, or after a transformation, or after cropping a volume, or even numerical error, streamlines end up having invalid coordinates (above the volume dimension or below 0 in voxel space). 
This will cause crash in function such as `target`, `density_map`, `connectivity_map`.

Removing such streamlines yourself can be surprisingly difficult, and since it is technically an invalid state for any tractogram. Having this function in the class will help user to save/load valid object.

PS : Loading and Saving will raise an error (which you can disable with an optional parameters) if trying to load/save an invalid stateful_tractogram.